### PR TITLE
tests: Add added base RPMI tests

### PR DESCRIPTION
**Pull request by slingappa**
Added Base service group tests.
Refer to test/README.md for mode details

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -102,6 +102,8 @@ ARFLAGS		=	rcs
 ELFFLAGS	+=	-static
 ELFFLAGS	+=	$(EXTRA_ELFFLAGS)
 
+LDFLAGS	+= -L$(build_dir) -lrpmi
```
Comment (avpatel) [inline Makefile:105]: Add these to `ELFFLAGS` just after `-static`

File: include/librpmi_env.h [hunk 2]
```diff
@@ -59,6 +59,32 @@ typedef long			rpmi_ssize_t;
...
 
+/**
+ * @brief compare memory from one place to another
+ *
+ * @param[in] s1	pointer to src1
+ * @param[in] s2	pointer to src2
+ * @param[in] n		number of bytes to compare
+ * @return void *	Pointer to destination
+ */
+static inline int rpmi_env_memcmp(void *s1, void *s2, rpmi_size_t n)
```
Comment (avpatel) [inline include/librpmi_env.h:70]: The change in librpmi_env.h should be a separate patch.

File: Makefile (hunk 1 above)
Comment (avpatel) [reply Makefile:105]: Also, the top-level Makefile change should be a separate patch.

File: test/test_shmem.c [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	/** Pointer to platform specific operations */
+	const struct rpmi_shmem_platform_ops *ops;
+	/** Private data of the platform specific oprations */
+	void *ops_priv;
+};
+
+static enum rpmi_error shmem_env_memcpy_read(void *priv, rpmi_uint64_t addr,
+					     void *in, rpmi_uint32_t len)
+{
+	rpmi_env_memcpy(in, (void *)addr, len);
```
Comment (avpatel) [inline test/test_shmem.c:36]: Never directly case integer to pointer. Instead first cast it to `unsigned long` and then cast it to pointer.

For example: `rpmi_env_memcpy(in, (void *)(unsigned long)addr, len);`

Comment (avpatel) [reply test/test_shmem.c:36]: This is to avoid compilation errors for 32bit systems.

File: Makefile (hunk 1 above)
Comment (slingappa) [reply Makefile]: Done

File: test/test_shmem.c (hunk 3 above)
Comment (slingappa) [reply test/test_shmem.c]: This file is gone now.

**Comment by slingappa** [comment]
@avpatel took care of the comments, please review

File: test/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+
+# 1. Overview
+
+This repo add RPMI tests for rpmi framework implemented on quemu
+
+# 2. Running tests
+
+## 2.1 Running qemu
+
+Execute qemu instance with shared memory backend for RAM:
```
Comment (avpatel) [inline test/README.md:10]: These tests are mock test which we can directly run as application on x86 host so no need to launch QEMU.

Comment (slingappa) [reply test/README.md:10]: can be yes, but some has to write the server if they're not using the qemu right ?

**Comment by avpatel** [comment]
Change the patch subject of last commit from `tests: Add added base RPMI tests` to `test: Add base RPMI tests`

File: test/README.md (hunk 4 above)
Comment (avpatel) [reply test/README.md:10]: better to have both client and server in the same test

Comment (avpatel) [reply test/README.md:10]: Basically each test should define the messages exchanged client and server. The test fails if expected response is not received a particular message.

Comment (avpatel) [reply test/README.md:10]: One test can be something like this:

1. Enqueue a request (message)
2. Process the context to get response
3. Dequeue the response (message)
4. Inspect the response to declare pass/fail

Comment (pathakraul) [reply test/README.md:10]: If possible, a summary of all tests passed/failed too

Comment (slingappa) [reply test/README.md:10]: These are already done:
- Basically each test should define the messages exchanged client and server. The test fails if expected response is not received a particular message.

- a summary of all tests passed/failed too, in the end , please refer to README.md inside test folder for sample output

**Comment by slingappa** [comment]
> test: Add base RPMI tests

done

File: test/README.md (hunk 4 above)
Comment (slingappa) [reply test/README.md:10]: > better to have both client and server in the same test

Server code inside librpmi would bloat it, as we'll have to paste all of the qemu RPMI implentation in the server code. better to point them to upstream qemu branch which already has all of it and provide samepl qemu command , like we have in the README file no ?

Comment (slingappa) [reply test/README.md:10]: > One test can be something like this:
> 
> 1. Enqueue a request (message)
> 2. Process the context to get response
> 3. Dequeue the response (message)
> 4. Inspect the response to declare pass/fail

This is done in _base* file, please check

Comment (avpatel) [reply test/README.md:10]: Okay, sounds good.

But, we still don't need to launch QEMU for running librpmi test applications.

Comment (slingappa) [reply test/README.md:10]: We do, for now base group might work with small snippet to initialize that, but for other groups like HSM, CPPC and all still needs to form the response another server app still needs to emulate, which essentially is being done in qemu.

If we dont want to do that, then, we'll have to do one of these:
1. Use the running qemu instance
2. write a test server app inside qemu code which links all the rpmi code
3. write a test server app out side qemu copy all the rpmi code.
So I still think option#1 above seems simplest one from users point of view.

Let me know if you guys would like to implement any one of teh above still or any other server emulation, will do it.

Comment (pathakraul) [reply test/README.md:10]: I thought that our current shmem(transport) can be entirely mocked on the x86 host? 
Both sides of transport (server and client) can be invoked in sequential manner to make one complete message request response transaction. Bigger test functions can be written which will comprise multiple such function calls in sequence of both sides(server and client)

I am just thinking from a user perspective where i dont want to have qemu launch which will require me to first build it and also find the right repo and things which go with it.

Thoughts?

Comment (slingappa) [reply test/README.md:10]: > I thought that our current shmem(transport) can be entirely mocked on the x86 host?
> 
> Both sides of transport (server and client) can be invoked in sequential manner to make one complete message request response transaction. Bigger test functions can be written which will comprise multiple such function calls in sequence of both sides(server and client)
> 
> I am just thinking from a user perspective where i dont want to have qemu launch which will require me to first build it and also find the right repo and things which go with it.
> 
> Thoughts?

ofcourse, it can be mocked, as mentioned in options 2 or 3 [here](https://github.com/riscv-software-src/librpmi/pull/19#discussion_r1690923413). 

But.. see my concerns mentioned in the same comment. if we really want to we can copy qemu implementation for current and future service groups. I personally dont think its a good idea to push an test implementation which bloats up librpmi.

**Comment by slingappa** [comment]
@avpatel @pathakraul ,
Added two more commits, which implements a test server to test Base group.

Now, no need to qemu, librpmi can be (mock)tested on any linux machine.

Again I'll let you guys to decide, 
- if and when we do decide this is what we need we can pick all these 5 commits,
- if we want to roll with qemu, then omit last 2 commits.

File: test/objects.mk [hunk 5]
```diff
@@ -3,13 +3,22 @@
 #
 # Copyright (c) 2024 Ventana Micro Systems Inc.
 #
+#
+test-elfs-y += test_srv
```
Comment (avpatel) [inline test/objects.mk:7]: I still don't understand. Why separate server application ?

With this approach, each test scenario will require a separate test application.

File: test/test_common.c [hunk 6]
```diff
@@ -5,3 +5,148 @@
...
+			hexdump("received",
+				(unsigned int *)msg->data,
+				rpmi_test->resp_data_len);
+			failed = 1;
+		}
+	}
+	printf("%-50s \t : %s!\n", rpmi_test->name, failed?"Failed":"Succeeded");
+}
+
+void execute_rpmi_test_group(struct rpmi_transport *rpmi_xport, struct rpmi_test_group *group)
```
Comment (avpatel) [inline test/test_common.c:77]: We can improve this common test infrastructure such that each test has both client and server part of the test-case. This will help us avoid creating a separate server application for each test scenario.

File: test/objects.mk (hunk 5 above)
Comment (slingappa) [reply test/objects.mk:7]: No, it doesnt, 

` process_ap_reqs()` inside `test_srv.c ` is common for all groups,
but server implementation definitely needs to be added just like we did in qemu for each service groups seperately as depicted below in  `add_*_group()` below:

```
 int main(int argc, char *argv[])
 {
         printf("Test server thread...\n");

         /*
          * along with contexts only BASE group is created, rest needs to be
          * added here later, after the context creation
          */
         create_contexts();
         
         add_hsm_group(&hsm);
         add_cppc_group(&cppc)
         
         process_ap_reqs();

         return 0;
 }

```

Comment (avpatel) [reply test/objects.mk:7]: This approach of adding a separate server for various scenarios is not going to scale. We have to re-design the common test infrastructure.

File: test/test_common.h [hunk 7]
```diff
@@ -7,4 +7,90 @@
...
+	int (*init)(struct rpmi_test *test);
+	int (*run)(struct rpmi_test *test);
+	int (*wait)(struct rpmi_test *test, struct rpmi_message *msg);
+	int (*verify)(struct rpmi_test *test, struct rpmi_message *msg);
+	int (*cleanup)(struct rpmi_test *test);
+};
+
+struct rpmi_test_group {
+	char name[64];
+	int num_tests;
```
Comment (avpatel) [inline test/test_common.h:70]: additonal members

rpmi_shmem *shmem;
rpmi_transport *trans;
rpmi_context *context;

File: test/test_common.h [hunk 8]
```diff
@@ -7,4 +7,90 @@
...
+	struct rpmi_test tests[];
+};
```
Comment (avpatel) [inline test/test_common.h:72]: setup()
cleanup()

**Comment by slingappa** [comment]
@avpatel @pathakraul ,

please check the changes now modified as per our discussion.

users are expected to modify `struct rpmi_test_group` members to create different test scenarios. and run `execute_test_group` repeatedly.

File: test/test_common.h [hunk 9]
```diff
@@ -7,4 +7,126 @@
...
+#include <stdio.h>
+#include <librpmi.h>
+#include <stdlib.h>
+
+/*
+ * Qemu shared memory offsets.
+ * its better to decode these data from DTS file, hardcoding for now.
+ */
+
+#define RPMI_SHM_XPORTS       (2)
```
Comment (avpatel) [inline test/test_common.h:19]: Drop this.

File: test/test_common.h [hunk 10]
```diff
@@ -7,4 +7,126 @@
...
+/*
+ * test assums qemu RPMI shared memory address is mapped at the following
+ * offset in gues RAM
+ */
+#define RPMI_SHM_OFSET       (0x0000000010001000)
```
Comment (avpatel) [inline test/test_common.h:24]: We don't need lot of these defines.

File: test/test_common.h [hunk 11]
```diff
@@ -7,4 +7,126 @@
...
+		rpmi_uint32_t hw_info_len;
+		const rpmi_uint8_t *hw_info;
+	} base;
+	char *priv;
+	int priv_len;
+};
+
+struct rpmi_test {
+	char name[64];
+	void *rpmi_xport;
```
Comment (avpatel) [inline test/test_common.h:71]: No need for this pointer. See below.

File: test/test_common.h [hunk 12]
```diff
@@ -7,4 +7,126 @@
...
+		unsigned int svc_id;
+		unsigned int svc_type;
+	} attrs;
+	int query_data_len;
+	unsigned char query_data[RPMI_MSG_MAX_DATA_SIZE];
+	int resp_data_len;
+	unsigned char resp_data[RPMI_MSG_MAX_DATA_SIZE];
+
+	/* test specific function handlers*/
+	int (*init)(struct rpmi_test *test);
```
Comment (avpatel) [inline test/test_common.h:85]: For all the test callbacks, just pass rpmi_test_group pointer which already has rpmi_context and rpmi_transport pointers.

File: test/test_transport.c [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+		rmsg->header.flags,
+		rmsg->header.servicegroup_id,
+		rmsg->header.service_id,
+		rmsg->header.datalen, rmsg->header.token);
+
+	for (i = 0; i < (rmsg->header.datalen / 4); i++)
+		printf("data[%2d]: %08x \n", i, data[i]);
+}
+
+void *get_shm_addr(unsigned int xport_id, unsigned int *shm_sz)
```
Comment (avpatel) [inline test/test_transport.c:35]: We don't need this anymore.

File: test/test_transport.c [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+	return calloc(1, (rpmi_xport->slot_size));
+}
+
+void free_rpmi_msg(struct rpmi_message *msg)
+{
+	if (msg)
+		free(msg);
+}
+
+struct rpmi_transport *init_test_rpmi_xport(int xport_id, int do_clear)
```
Comment (avpatel) [inline test/test_transport.c:114]: This function is not need. Most of this things can be done in init_group().

In fact, you can move important stuff from test_transport.c and remove test_transport.c

File: test/test_common.c [hunk 15]
```diff
@@ -5,3 +5,237 @@
...
+			group->init(group);
+	}
+
+	return 0;
+}
+
+void cleanup_group(struct rpmi_test_group *group)
+{
+	free(group->xport->priv);
+	free(group->rctx);
```
Comment (avpatel) [inline test/test_common.c:208]: Use librpmi APIs.

File: test/test_common.h (hunk 9 above)
Comment (slingappa) [reply test/test_common.h:19]: agreed, dropping.

File: test/test_common.h (hunk 10 above)
Comment (slingappa) [reply test/test_common.h:24]: agreed, was needed for qemu, not anymore dropping.

File: test/test_common.h (hunk 12 above)
Comment (slingappa) [reply test/test_common.h:85]: No, rpmi_test still needs as we still need to track, request/response data per service group tests.

**Comment by slingappa** [comment]
@avpatel 
- `test_common.c` allocates/frees for every test memory instead of opening `/dev/shm/` file
- removed `*transport.c` file and brought in required changes inside `common.c` file
- removed multi transport related changes, so `xport_id` gone from everywhere.

File: include/librpmi_env.h [hunk 16]
```diff
@@ -368,7 +394,10 @@ static inline rpmi_uint32_t rpmi_to_xe32(rpmi_bool_t is_be, rpmi_uint32_t val)
  * @param[in] addr	64-bit address
  * @param[in] val	4-byte data to write
  */
-void rpmi_env_writel(rpmi_uint64_t addr, rpmi_uint32_t val);
+static inline void rpmi_env_writel(rpmi_uint64_t addr, rpmi_uint32_t val)
```
Comment (avpatel) [inline include/librpmi_env.h:397]: This should be implemented in test_common.c

**Comment by avpatel** [comment]
@slingappa Much better now. Few comments remaining.

File: test/test_common.c (hunk 15 above)
Comment (slingappa) [reply test/test_common.c]: Done
